Implement profiling for compiler-generated move/copy operations (MIR transform)

jsgf opened the pull request:
As part of Rust's move semantics, the compiler will generate memory copy operations to move objects about. These are generally pretty small, and the backend is good at optimizing them. But sometimes, if the type is large, they can end up being surprisingly expensive. In such cases, you might want to pass them by reference, or Box them up.

However, these moves are also invisible to profiling. At best they appear as a `memcpy`, but one memcpy is basically indistinguishable from another, and its very hard to know that 1) it's actually a compiler-generated copy, and 2) what type it pertains to.

This PR adds two new pseudo-intrinsic functions in `core::intrinsics`:
```
pub fn compiler_move<T, const SIZE: usize>(_src: *const T, _dst: *mut T);
pub fn compiler_copy<T, const SIZE: usize>(_src: *const T, _dst: *mut T);
```
These functions are never actually called however. A MIR transform pass -- `instrument_moves.rs` -- will locate all `Operand::Move`/`Copy` operations, and modify their source location to make them appear as if they had been inlined from `compiler_move`/`_copy`.

These functions have two generic parameters: the type being copied, and its size in bytes. This should make it very easy to identify which types are being expensive in your program (both in aggregate, and at specific hotspots). The size isn't strictly necessary since you can derive it from the type, but it's small and it makes it easier to understand what you're looking at.

This functionality is only enabled if you have debug info generation enabled, and also set the `-Zinstrument-moves` option.

It does not instrument all moves. By default it will only annotate ones for types over 64 bytes. The `-Zinstrument-moves-size-limit` specifies the size in bytes to start instrumenting for.

This has minimal impact on the size of debugging info. For rustc itself, the overall increase in librustc_driver*.so size is around .05% for 65 byte limit, 0.004% for 1025 byte limit, and a worst case of 0.6% for an 8 byte limit.

There's no effect on generated code, it only adds debug info.

As an example of a backtrace:
```
Breakpoint 1.3, __memcpy_avx512_unaligned_erms () at ../sysdeps/x86_64/multiarch/memmove-vec-unaligned-erms.S:255
255	ENTRY_P2ALIGN (MEMMOVE_SYMBOL (__memmove, unaligned_erms), 6)
(gdb) bt
 # 0  __memcpy_avx512_unaligned_erms () at ../sysdeps/x86_64/multiarch/memmove-vec-unaligned-erms.S:255
 # 1  0x0000555555590e7e in core::intrinsics::compiler_copy<[u64; 1000], 8000> () at library/core/src/intrinsics/mod.rs:10
 # 2  t::main () at t.rs:10
```

<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r? <reviewer name>
-->
<!-- homu-ignore:end -->

jsgf commented:
I'm not really sure whether Statement::Assign's rvalues covers all the interesting cases or not. I'd like to make sure it covers:
- parameter passing
- returns
- assignment
- initialization
- ...anything else?

rust-log-analyzer commented:
The job **`aarch64-gnu-llvm-20-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/18149237234/job/51656721695) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/51656721695) [(plain)](https://github.com/rust-lang/rust/commit/5cc27d43d409dd1c35092f734dfe2dbae4c837b9/checks/51656721695/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s


running 2 tests
test /checkout/src/doc/unstable-book/src/compiler-flags/instrument-moves.md - The_::Behavior (line 45) ... FAILED
test /checkout/src/doc/unstable-book/src/compiler-flags/instrument-moves.md - The_::Example (line 53) ... ok

failures:

---- /checkout/src/doc/unstable-book/src/compiler-flags/instrument-moves.md - The_::Behavior (line 45) stdout ----
error: expected item, found `0`
##[error] --> /checkout/src/doc/unstable-book/src/compiler-flags/instrument-moves.md:46:1
  |
1 | 0: memcpy
  | ^ expected item
---

Couldn't compile the test.

failures:
    /checkout/src/doc/unstable-book/src/compiler-flags/instrument-moves.md - The_::Behavior (line 45)

test result: FAILED. 1 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.13s

Command `/checkout/obj/build/bootstrap/debug/rustdoc -Wrustdoc::invalid_codeblock_attributes -Dwarnings -Znormalize-docs -Z unstable-options --test /checkout/src/doc/unstable-book/src/compiler-flags/instrument-moves.md --test-args ` failed with exit code 101
Created at: src/bootstrap/src/core/builder/mod.rs:1682:23
Executed at: src/bootstrap/src/core/build_steps/test.rs:2694:13

Command has failed. Rerun with -v to see more details.
Bootstrap failed while executing `--stage 2 test --skip tests --skip coverage-map --skip coverage-run --skip library --skip tidyselftest`
Build completed unsuccessfully in 0:38:38
  local time: Wed Oct  1 03:07:37 UTC 2025
  network time: Wed, 01 Oct 2025 03:07:38 GMT
##[error]Process completed with exit code 1.
```

</details>

RalfJung commented on library/core/src/intrinsics/mod.rs:
```diff
@@ -3314,3 +3314,27 @@ pub unsafe fn va_arg<T: VaArgSafe>(ap: &mut VaListImpl<'_>) -> T;
 #[rustc_nounwind]
 pub unsafe fn va_end(ap: &mut VaListImpl<'_>);
+
+/// Compiler-generated move operation - never actually called.
+/// Used solely for profiling and debugging visibility.
+///
+/// This function serves as a symbolic marker that appears in stack traces
+/// when rustc generates move operations, making them visible in profilers.
+/// The SIZE parameter encodes the size of the type being moved in the function name.
+#[rustc_force_inline]
+#[rustc_diagnostic_item = "compiler_move"]
+pub fn compiler_move<T, const SIZE: usize>(_src: *const T, _dst: *mut T) {
+    unreachable!("compiler_move should never be called - it's only for debug info")
+}
+
+/// Compiler-generated copy operation - never actually called.
+/// Used solely for profiling and debugging visibility.
+///
+/// This function serves as a symbolic marker that appears in stack traces
+/// when rustc generates copy operations, making them visible in profilers.
+/// The SIZE parameter encodes the size of the type being copied in the function name.
+#[rustc_force_inline]
+#[rustc_diagnostic_item = "compiler_copy"]
+pub fn compiler_copy<T, const SIZE: usize>(_src: *const T, _dst: *mut T) {
+    unreachable!("compiler_copy should never be called - it's only for debug info")
+}
```
These aren't intrinsics so I don't think this is the best place for them. The file is already too big anyway.^^

jsgf replied:
Yeah I wasn't sure exactly where to put them. Originally I had the idea of actually making them real functions implementing copy & move in terms of calls to them, but that seemed more fiddly than it's worth.

jsgf commented:
> Is it really worth distinguishing moves and copies? That doesn't make much of a difference for the runtime code, it's mostly a type system thing.

I think in practice it's useful - I've seen very large structures being made `Copy` just because all their fields allow it and then being copied around unexpectedly. It would be nice to be able to see that, and distinguish it from regular Move.

> The codegen backend would have an easier time centralizing all the ways in which operand uses get codegen'd as memcpy. But I am not sure if there's still a good way to adjust debuginfo there...

That was actually my first attempt but I ended up with a stream of mysterious crashes/assertion failures from within the guts of llvm. Doing the manipulations at the MIR level turned out to be much more straightforward.

jsgf commented:
@RalfJung 
- I created core::profiling to hold compiler_copy/compiler_move
- I looked at the `visit_operand` visitor; it's appealing but it doesn't give access to change the containing source info.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-20`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/18172760999/job/51731331335) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/51731331335) [(plain)](https://github.com/rust-lang/rust/commit/2ba9b47ae8ff6c2409a56f24c239a42606e5e1a3/checks/51731331335/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
test [mir-opt] tests/mir-opt/unusual_item_types.rs ... ok

failures:

---- [mir-opt] tests/mir-opt/instrument-moves/iter.rs stdout ----
7         debug iter => _1;
8         scope 2 (inlined test_impl_trait_arg::<std::iter::Chain<Map<std::array::IntoIter<u64, 5>, fn(u64) -> u64>, Map<std::array::IntoIter<u64, 5>, fn(u64) -> u64>>>) {
9             debug iter => _1;
-             scope 3 (inlined core::profiling::compiler_move::<std::iter::Chain<Map<std::array::IntoIter<u64, 5>, fn(u64) -> u64>, Map<std::array::IntoIter<u64, 5>, fn(u64) -> u64>>, 128>) {
+             scope 3 (inlined core::profiling::compiler_move::<std::iter::Chain<Map<std::array::IntoIter<u64, 5>, fn(u64) -> u64>, Map<std::array::IntoIter<u64, 5>, fn(u64) -> u64>>, 104>) {
11             }
12         }
13     }


```

</details>

saethlin commented:
All codegen tests have this bittleness. We build all codegen tests (mir-opt, codegen, assembly) suites against a sysroot which is compiled with whatever flags are set in the user-provided profile. The mir-opt suite tends to get blamed because we check in much more of the MIR than just the FileCheck annotations.

I suspect your test needs a `//@ ignore-std-debug-assertions`. The particular diff above is because the inliner changed behavior because the debug assertions changed the inlining cost, but even if you stabilized the inliner the MIR would depend on debug assertions.

RalfJung commented:
Is "instrument" the best term here? Usually that means to actually run some extra code for the to-be-instrumented operation, doesn't it? This here is just adding debuginfo.

Do new `-Z` flags need MCP? Judging from https://forge.rust-lang.org/compiler/proposals-and-stabilization.html#compiler-flags and assuming that this is meant not just for internal use by rustc developers, I think the answer is "yes".

nnethercote commented:
Can `-Zinstrument-moves` and `-Zinstrument-moves-size-limit` be combined?

How can the added debuginfo be used to produce useful profiling information?

jsgf commented:
I ended up renaming it to `--annotate-moves` and using a single option:
- `--annotate-moves` - enable with default limit
- `--annotate-moves=true/false/on/off` explicitly enable/disable
- `--annotate-moves=1024` enable with a specific size limit

jsgf committed:
Add -Zannotate-moves for profiler visibility of move/copy operations

This implements a new unstable compiler flag `-Zannotate-moves` that makes
move and copy operations visible in profilers by creating synthetic debug
information. This is achieved with zero runtime cost by manipulating debug
info scopes to make moves/copies appear as calls to `compiler_move<T, SIZE>`
and `compiler_copy<T, SIZE>` marker functions in profiling tools.

A new `AnnotateMoves` MIR transform pass runs after MIR optimization
and modifies source scopes for statements containing `Operand::Move`
and `Operand::Copy` to make them appear as if inlined from profiling
marker functions in `core::profiling`.

Two marker functions (`compiler_move` and `compiler_copy`) are defined
in `library/core/src/profiling.rs`. These are never actually called -
they exist solely as debug info anchors.

The transform creates synthetic `SourceScopeData` with the `inlined`
field set to point to the appropriate profiling marker, leveraging the
existing inlining infrastructure.

Operations are only annotated if the type:
   - Meets the size threshold (default: 65 bytes, configurable via
     `-Zannotate-moves=SIZE`)
   - Has a non-scalar backend representation (scalars use registers,
   not memcpy)

An early issue was that modifying a statement's SourceInfo to add the
`compiler_move` scope affected the entire statement, including function calls
when the move was a call argument. This made profilers attribute the whole
function call to the move operation, greatly exaggerating its cost.

The solution stores argument move/copy SourceInfo separately in
`Body::call_arg_move_source_info`. During codegen, this SourceInfo is applied
only during argument preparation, then reset to the call site location before
emitting the call instruction itself. This ensures profilers see the argument
copy attributed to `compiler_move` while the function call retains its proper
source attribution.

jsgf committed:
Put move source locations for parameters into Call/TailCall

Rather than store parameter moves in a separate side structure, which
relies on MIR indexes not changing again, instead move the parameter
move source_info into TerminatorKind::Call/TailCall itself. This keeps
everything together as a parallel array of parameter move information.

If we aren't annotating moves, or no parameters need annotation, then
then remains as None, Otherwise it's a `Box<[Option<SourceInfo>]>`
recording the source location of the annotated parameters.

This ends up touching a lot of files with mostly one line changes just
because of the extra field in the Call/TailCall enums. I changed a lot
of the patterns to use `..` wildcard matching since in many cases most
fields were being skipped.

jsgf commented:
I added a commit to put the parameter move info into `TerminatorKind::Call`/`TailCall`. It's semantically simple, but has a pretty large blast-radius of trivial changes because of the type change.

jsgf commented:
I revisited doing these annotations closer to codegen, and I think it came out simpler and nicer. 

https://github.com/rust-lang/rust/pull/147803
